tools/ami-cleanup/src/cleanup/cleanup.go

```diff
@@ -0,0 +1,237 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cleanup
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/account"
+	accounttypes "github.com/aws/aws-sdk-go-v2/service/account/types"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/gravitational/trace"
+	"github.com/relvacode/iso8601"
+)
+
+const maxImageKeepAge = 30 * 24 * time.Hour // Images older than this will be removed, if they match other required criteria.
+
+type ApplicationInstance struct {
+	shouldDoDryRun bool
+
+	// Dependency injection for testing
+	accountClientGenerator func(cfg *aws.Config) AccountAPI
+	ec2ClientGenerator     func(cfg *aws.Config) EC2API
+}
+
+// NewApplicationInstance creates a new instance of the tool
+func NewApplicationInstance(doDryRun bool) *ApplicationInstance {
+	return &ApplicationInstance{
+		shouldDoDryRun:         doDryRun,
+		accountClientGenerator: NewAccountAPI,
+		ec2ClientGenerator:     NewEC2API,
+	}
+}
+
+// Run performs cleanup of AMIs in all enabled regions.
+func (ai *ApplicationInstance) Run(ctx context.Context) error {
+	cfg, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return trace.Wrap(err, "failed to load AWS credentials")
+	}
+
+	enabledRegions, err := ai.getEnabledRegions(ctx, ai.accountClientGenerator(&cfg))
+	if err != nil {
+		return trace.Wrap(err, "failed to get enabled regions")
+	}
+
+	if len(enabledRegions) == 0 {
+		return nil
+	}
+
+	totalBytesRecovered := int32(0)
+	totalImagesDeleted := 0
+	// This would probably run much faster with channels/concurrency, but there isn't much
+	// need with how often this is expected to be ran
+	for _, enabledRegion := range enabledRegions {
+		spaceRecovered, imagesDeleted, err := ai.cleanupRegion(ctx, cfg, *enabledRegion.RegionName)
+		if err != nil {
+			return trace.Wrap(err, "failed to clean up images in region %q", *enabledRegion.RegionName)
+		}
+
+		totalBytesRecovered += spaceRecovered
+		totalImagesDeleted += imagesDeleted
+	}
+
+	log.Printf("Deleted %d GiB of %d images across %d regions", totalBytesRecovered, totalImagesDeleted, len(enabledRegions))
+	return nil
+}
+
+// Cleans up the images in the given region. Returns the total amount of space recovered, number of images deleted,
+// and any error.
+func (ai ApplicationInstance) cleanupRegion(ctx context.Context, cfg aws.Config, regionName string) (int32, int, error) {
```

Comment:
```suggestion
func (ai *ApplicationInstance) cleanupRegion(ctx context.Context, cfg aws.Config, regionName string) (int32, int, error) {
```

It's a bit odd to have a mix of pointer and non-pointer receivers on the same type.